qemu/tests/qemu_guest_agent.py

```diff
@@ -512,6 +512,131 @@ def gagent_check_set_vcpus(self, test, params, env):
         if vcpus_info[vcpus_num - 1]["online"] is not False:
             test.fail("the vcpu status is not changed as expected")
 
+    @error_context.context_aware
+    def gagent_check_set_mem_blocks(self, test, params, env):
+        """
+        Get/set logical memory blocks via guest agent.
+        Steps:
+        1) Get the size of memory block unit via guest agent
+        2) Offline one memory block which can be removable in guest
+        3) Verify memory blocks via guest agent is offline
+        4) Verify memory block unit size
+        5) Offline some memory blocks which can be offline via guest agent
+        6) Verify memory are decreased in guest
+        7) Online the memory blocks which are offline before
+        8) Verify memory are the same as before
+        9) Offline a memroy block which can't be offline
+
+        :param test: kvm test object
+        :param params: Dictionary with the test parameters
+        :param env: Dictionary with test environment.
+        """
+        session = self._get_session(params, None)
+        self._open_session_list.append(session)
+        cmd_get_mem = "free -m |grep -i mem"
+        cmd_offline_mem = "echo 0 > /sys/devices/system/memory/memory%s/online"
+        # record the memory blocks phys-index which is set to offline
+        mem_phys_index_list = []
+
+        error_context.context("Get the size of memory block unit.",
+                              logging.info)
+        mem_block_info = self.gagent.get_memory_block_info()["size"]
+        mem_unit_size = mem_block_info / float(1024 * 1024)
+
+        error_context.context("Offline one memory block in guest.",
+                              logging.info)
+        mem_size_original = session.cmd_output(cmd_get_mem).strip().split()[1]
+        mem_blocks = self.gagent.get_memory_blocks()
+        list_index = 0
+        for memory in mem_blocks:
+            if memory["online"] and memory["can-offline"]:
+                mem_phys_index = memory["phys-index"]
+                mem_phys_index_list.append(mem_phys_index)
+                break
+            list_index += 1
+        else:
+            logging.info("All memory blocks are offline already.")
+            return
+        session.cmd(cmd_offline_mem % mem_phys_index)
+
+        error_context.context("Verify it's changed to offline status via"
+                              " agent.", logging.info)
+        mem_blocks = self.gagent.get_memory_blocks()
+        if mem_blocks[list_index]["online"] is not False:
+            test.fail("%s phys-index memory block is still online"
+                      " via agent." % mem_phys_index)
+
+        error_context.context("Verify the memory block unit size.",
+                              logging.info)
+        mem_size = session.cmd_output(cmd_get_mem)
+        mem_size_aft_offline_guest = mem_size.strip().split()[1]
+        delta = float(mem_size_original) - float(mem_size_aft_offline_guest)
+        if delta != mem_unit_size:
```

Comment:
@xiagao If offline smaller memory size.  maybe delta != mem_unit_size,

Reply:
> @xiagao If offline smaller memory size. maybe delta != mem_unit_size,

I only offline one memory block, and the diff is just the block unit.